snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/LinkDecoratorTest.kt

```diff
@@ -0,0 +1,179 @@
+        val result = tracker.decorateLink(testLink)
+
+        matchesRegex(pattern, result!!)
+    }
+
+    @Test
+    fun testDecorateUriWithExistingSpParam() {
+        tracker.track(ScreenView("test"))
+
+        val pattern =
+            Regex("""http://example\.com\?_sp=$userId\.\d{13}\.${session.sessionId}\.decoratorTest\.mob\.subjectUserId""")
+        val result = tracker.decorateLink(testLink.buildUpon().appendQueryParameter("_sp", "test").build())
+
+        matchesRegex(pattern, result!!)
+    }
+
+    @Test
+    fun testDecorateUriWithOtherParam() {
+        tracker.track(ScreenView("test"))
+
+        val pattern =
+            Regex("""http://example\.com\?a=b&_sp=$userId\.\d{13}\.${session.sessionId}\.decoratorTest\.mob\.subjectUserId""")
+        val result =
+            tracker.decorateLink(testLink.buildUpon().appendQueryParameter("a", "b").build())
+
+        matchesRegex(pattern, result!!)
+    }
+
+    @Test
+    fun testDecorateUriWithParameters() {
+        tracker.track(ScreenView("test"))
+
+        val sessionId = session.sessionId
+        val expectedParams = hashMapOf(
+            listOf(CrossDeviceParameter.SESSION_ID) to ".$sessionId",
+
+            listOf(
+                CrossDeviceParameter.SESSION_ID,
+                CrossDeviceParameter.SOURCE_ID
+            ) to ".$sessionId.decoratorTest",
+
+            listOf(
+                CrossDeviceParameter.SESSION_ID,
+                CrossDeviceParameter.SOURCE_ID,
+                CrossDeviceParameter.SOURCE_PLATFORM
+            ) to ".$sessionId.decoratorTest.mob",
+
+            listOf(
+                CrossDeviceParameter.SESSION_ID,
+                CrossDeviceParameter.SOURCE_ID,
+                CrossDeviceParameter.SOURCE_PLATFORM,
+                CrossDeviceParameter.USER_ID
+            ) to ".$sessionId.decoratorTest.mob.subjectUserId",
+
+            listOf(
+                CrossDeviceParameter.SESSION_ID,
+                CrossDeviceParameter.SOURCE_PLATFORM,
+            ) to ".$sessionId..mob",
+
+            listOf(
+                CrossDeviceParameter.SOURCE_ID,
+                CrossDeviceParameter.USER_ID
+            ) to "..decoratorTest..subjectUserId",
+
+            listOf(
+                CrossDeviceParameter.USER_ID
+            ) to "....subjectUserId",
+
+            emptyList<CrossDeviceParameter>() to "",
+        )
+
+        for ((param, spVal) in expectedParams.entries) {
+            val pattern =
+                Regex("""http://example\.com\?_sp=$userId\.\d{13}$spVal""")
+            val result = tracker.decorateLink(testLink, param)
+
+            matchesRegex(pattern, result!!)
+        }
+    }
+
+    fun getTracker(): TrackerController {
+        val context = InstrumentationRegistry.getInstrumentation().targetContext
+
+        val networkConfiguration = NetworkConfiguration("fake-url", HttpMethod.POST)
```

Comment:
We prefer to use the mocked network connection in tests so that we don't make unnecessary requests, see here: https://github.com/snowplow/snowplow-android-tracker/blob/master/snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/PluginsTest.kt#L219